scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+        total_grad_norm_constraint: float = 200,
+        use_gpu: Optional[bool] = None,
+        train_size: float = 1,
+        validation_size: Optional[float] = None,
+        batch_size = None,
+        plan_kwargs: Optional[dict] = None,
+        **kwargs,
+    ):
+        """
+        Train the model using variational inference (Pyro).
+
+        Parameters
+        ----------
+        max_epochs
+            Number of epochs to train for
+        lr
+            Learning rate for optimization.
+        use_gpu
+            If `True`, use the GPU if available.
+        train_size
+            Size of training set in the range [0.0, 1.0].
+        validation_size
+            Size of the test set. If `None`, defaults to 1 - `train_size`. If
+            `train_size + validation_size < 1`, the remaining cells belong to a test set.
+        batch_size
+            Minibatch size to use during training.
+        **kwargs
+            Other keyword args for :class:`~scvi.lightning.Trainer`.
+        """
+        
+        if use_gpu is None:
+            use_gpu = self.use_gpu
+        else:
+            use_gpu = use_gpu and torch.cuda.is_available()
+        gpus = 1 if use_gpu else None
+        
+        if self.use_gpu:
+            torch.set_default_tensor_type(torch.cuda.FloatTensor)
+        else:
+            torch.set_default_tensor_type(torch.FloatTensor)
+        
+        
+        if use_gpu:
+            self.module.cuda()
+        
+        train_dl = AnnDataLoader(self.adata, shuffle=True, 
+                                 batch_size=self.batch_size, 
+                                 drop_last=True)
+        pyro.clear_param_store()
+        module = self.module
+        # warmup guide for JIT
+        for tensors in train_dl:
+            args, kwargs = module._get_fn_args_from_batch(tensors)
+            module.guide(*args, **kwargs)
+            break
+            
+        train_dl = AnnDataLoader(self.adata, shuffle=True, 
+                                 batch_size=self.batch_size, 
+                                 drop_last=True)
+        plan = PyroTrainingPlan(module,
+                                loss_fn=pyro.infer.Trace_ELBO(), 
+                                # for some reason JitTrace_ELBO does not work (AssertionError)
+                                optim=pyro.optim.ClippedAdam({'lr': lr,
+                                                        # limit the gradient step from becoming too large
+                                                        'clip_norm': total_grad_norm_constraint}))
+        self.trainer = Trainer(
+            gpus=gpus,
+            max_epochs=max_epochs,
+        )
+        self.trainer.fit(plan, train_dl)
+        
+        #try:
+        #    self.history = self.trainer.logger.history
+        #except AttributeError:
+        #    self.history = None
+
+        self.is_trained_ = True
+
+        
+    def sample_node1(self, node, num_samples_batch: int = 10):
```

Comment:
1. This function takes one batch of samples (e.g. 10 samples) from the posterior distribution. To generate 1000 samples, `sample_node1` with `num_samples_batch=10` is applied 100 times here: https://github.com/YosefLab/scvi-tools/pull/930/files/6aaf614765a89fa1b3da5689b83e354ab44aa05a#diff-f59806d7a4b2b195cb84ed4b7bbe89a982a0f3cea4819904e413f685627b593fR221-R235

2. On using scVI data loader as you suggest: 
A) Concatenating local parameters that come out of each data loader batch is non-trivial. It probably needs digging out which parameters belong to a batch plate (which means that batch plate should always have the same name, e.g.  `obs_plate`), then concatenating those parameters in the axis of that plate.

B) Global parameters will be sampled in every dataloader batch essentially meaning that while local parameters will have 1000 samples, global parameters will have `1000 * n_obs / batch_size` samples. I am not sure how to deal with this ... use only the last/first dataloader batch?

C) Doing a pass through data with dataloader (and repeating that 100 times to get 100 * 10 posterior samples) can be slow because it would move X to the GPU too many times. Asking for 1000 samples at the same time will likely hit GPU memory when there are many global parameters.

D) Unless guide is performing the autoencoding operation, X and other observed variables are actually not used by Predictive. 

E) Given D, it seems sensible to ignore dataloader batches and just give the model indices of all observations. We also do not want to move all observations in X to GPU so it seems sensible to replace it with arbitrary tensor.

Would be good if scVI provides general methods for doing Predictive for pyro models because this step is quite model agnosic. 

What do you think?